mu/virtual_environment.py

```diff
@@ -0,0 +1,553 @@
+
+    def install_jupyter_kernel(self):
+        logger.info("Installing Jupyter Kernel")
+        return self.run_python(
+            "-m",
+            "ipykernel",
+            "install",
+            "--user",
+            "--name",
+            self.name,
+            "--display-name",
+            '"Python/Mu ({})"'.format(self.name),
+        )
+
+    def install_baseline_packages(self):
+        """Install all packages needed for non-core activity
+
+        Each mode needs one or more packages to be able to run: pygame zero
+        mode needs pgzero and its dependencies; web mode needs Flask and so on.
+        We intend to ship with all the necessary wheels for those packages so
+        no network access is needed. But if the wheels aren't found, because
+        we're not running from an installer, then just pip install in the
+        usual way.
+
+        --upgrade is currently used with a thought to upgrade-releases of Mu
+        """
+        logger.info("Installing baseline packages")
+        logger.info(
+            "%s %s",
+            wheels_dirpath,
+            "exists" if os.path.isdir(wheels_dirpath) else "does not exist",
+        )
+        #
+        # This command should install the baseline packages, picking up the
+        # precompiled wheels from the wheels path
+        #
+        # For dev purposes (where we might not have the wheels) warn where
+        # the wheels are not already present and download them
+        #
+        wheel_filepaths = glob.glob(os.path.join(wheels_dirpath, "*.whl"))
+        if not wheel_filepaths:
+            logger.warn(
+                "No wheels found in %s; downloading...", wheels_dirpath
+            )
+            wheels.download()
+            wheel_filepaths = glob.glob(os.path.join(wheels_dirpath, "*.whl"))
+
+        if not wheel_filepaths:
+            raise RuntimeError(
+                "No wheels in %s; try `python -mmu.wheels`" % wheels_dirpath
+            )
+        self.pip.install(wheel_filepaths)
+
+    def register_baseline_packages(self):
+        """Keep track of the baseline packages installed into the empty venv"""
+        #
+        # FIXME: This should go into settings.
+        # For now, though, just put it somewhere
+        #
+        packages = list(self.pip.installed())
+        os.makedirs(
+            os.path.dirname(self.BASELINE_PACKAGES_FILEPATH), exist_ok=True
+        )
+        with open(self.BASELINE_PACKAGES_FILEPATH, "w", encoding="utf-8") as f:
+            json.dump(packages, f)
+
+    def baseline_packages(self):
+        """Return the list of baseline packages"""
+        #
+        # FIXME: This should come out of settings. For now though...
+        #
+        with open(self.BASELINE_PACKAGES_FILEPATH, encoding="utf-8") as f:
+            try:
+                return json.load(f)
+            except json.decoder.JSONDecodeError:
+                logger.exception("Unable to read baseline packages")
+                return []

```

Comment:
Perhaps we should already when merging create an issue for this, I think we have various other issues related to the config, would make sense to fix it while looking at those.